includes/components/class-attribute.php

```diff
@@ -85,9 +85,138 @@ protected function boot() {
 
 			// Disable Jetpack search.
 			add_filter( 'jetpack_search_should_handle_query', [ $this, 'disable_jetpack_search' ], 10, 2 );
+
+			// Set search order.
+			add_filter( 'posts_orderby', [ $this, 'set_search_order' ] );
+		} else {
+
+			// Alter settings.
+			add_filter( 'hivepress/v1/settings', [ $this, 'alter_settings' ] );
 		}
 	}
 
+	/**
+	 * Sets WP search order.
+	 *
+	 * @param string $orderby Order filters.
+	 * @return string
+	 */
+	public function set_search_order( $orderby ) {
```

Comment:
Let me know why a separate callback was required, we already have a function which validates the sort form in the set_search_query (it also already has conditions for attribute-model queries), if possible move the code there

Reply:
Please also make sure that the code is optimized and is as simple as possible

Reply:
It's not working in the `set_search_query` method because check for model is not passing

Reply:
I meant this part https://github.com/hivepress/hivepress/blob/master/includes/components/class-attribute.php#L1761 Before this part of the code the query is already checked (that it's the main query of the attribute model, so it's applied to e.g. Listings page and listing search).

Also, I realized that another approach may work - if you simply move the option which is selected as default in Settings to the top of the sort form drop-down (via the form filter callback), and make sure that it's selected by default (e.g. if you open the Listings page), then no extra code changes may be required and it may start working because this code:

`$sort_form->get_value( '_sort' );`

Will get the correct value by default, and the current logic will be applied which already sorts listings here https://github.com/hivepress/hivepress/blob/master/includes/components/class-attribute.php#L1761

Reply:
Please check why it's empty, if this was the only reason then altering this condition is a more simple solution (with the aim to re-use the existing logic which was available) than implementing a similar logic on the SQL level. This new "default sorting attribute" parameter should work for the following cases:

- the Listings page
- the search page
- category (and other taxonomy, e.g. tag) pages

The second one is covered by the is_post_type_archive check, the third one by is_tax (before the model name check on the line you mentioned), do you mean that the first case doesn't work (the Listings page)? Then we can add another check to also include the Listings page query somehow. 

Then the sorting-related code will run on the Listings page and in 2 other cases and we'll be able to re-use it, by just setting the default selected drop-down value in the form.

Reply:
Also make sure that this works for all the searchable attribute-enabled models, avoid using hard-coded list of models, I remember discussing to use the "searchable" flag in the models array config in the class parameter (you can set it via a separate callback or in the boot method, by checking the "has_archive" parameter of the post type config of each model). A similar check is already available in Attribute component, we used it to show/hide the Search meta box of the attribute settings, if you use a dedicated callback we can then simply check this like if($this->models[$model]['searchable'])...

Reply:
I've already described why $model is empty in the discussion above, I will repeat it here:

> из-за того, что `pre_get_posts` срабатывает слишком рано (до инициализации страницы), то очень сложно отследить модель и ухватиться хотя бы за какие-нибудь параметры чтобы можно было про-ассоциировать страницу с моделью (пробовал `is_post_type_archive`, `is_archive`, `global $post`, `get_queried_object` и пр.)

What about hard-coded list of models: I've already implemented `get_searchable_models` method in the previous commit, so we don't using hard-coded models anymore.

Reply:
I found a workaround for this. The query wasn't applied to the Listings page because the $query->is_main_query() failed, not the model name one. This because the main query is the page itself, and the listings query is a secondary one, we pass it via the "posts_query" request context parameter and the query occurs in the Page block rendering.

I added a custom flag to the query parameters (check the changes in class-page.php above) to be able to differentiate these custom "main" queries, now the set_search_query callback is applied to all 3 cases (the Listings page, search pages, category pages) so you can proceed with re-using the existing sort logic. 

Please wrap other code within the callback with "!$query->get('hp_main')" condition (all parts except the sorting one and the $query->is_search() one because this condition will not pass on the Listings page anyway), this way the sorting logic will be applied to these "hp_main" queries but we'll ensure that other parts of the callback will not affect it (since these were not applied to the "hp_main" queries previously it's better to wrap them).

Regarding the get_searchable_models helper, please create a callback instead as suggested because the main reason for this is to run the code once. If you use a helper the array filtering will run every time you call it, with the callback it runs once during the HTTP request and will basically "cache" the searchable parameter in the models config array. Then you can simply check if($this->models[$model]['searchable'])... Please also avoid creating public helpers, because third-party developers may start using them from the outside code and it'll be hard to deprecate the code without breaking compatibility, and we can change any implementation this way (e.g. callbacks, function names, splitting functions) while keeping the interface (e.g. "searchable" flag which is unlikely to change) the same.

Reply:
Adding new `hp_main` flag nothing changed on my side
I've debugged `$query->is_main_query()` check and it's always `true` when I'm on the archive pages (e.g. listings, vendors, requests), so the problem is still in `$model`, it's always `null`